File: tests/live/target.py
"""Debug target for exercising the bridge: globals, nested frames, locals."""

import sys

g_connection_count = 17
g_hostname = "prod-db-01"
g_config = {"retries": 3, "timeout_ms": 5000, "tags": ["a", "b"]}


class Header:
    def __init__(self, length, magic):
        self.length = length
        self.magic = magic

    def __repr__(self):
        return f"Header(length={self.length}, magic={self.magic:#x})"


def parse_header(raw, limit):
    header = Header(len(raw), 0xDEADBEEF)
    scratch = [i * i for i in range(6)]
    total = sum(scratch)
    if header.length > limit:
        raise ValueError(f"header too long: {header.length} > {limit}")
    return header, total


def handle_packet(payload, limit):
    print(f"handling {len(payload)} bytes", flush=True)
    header, total = parse_header(payload, limit)
    return header, total


def main():
    print("target starting", flush=True)
    header, total = handle_packet(b"hello world", 4096)
    print(f"ok: {header} total={total}", flush=True)
    print(f"globals: {g_connection_count} {g_hostname} {g_config}", flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/live/target.cpp
// Live-test target: globals, nested frames, a struct, a byte buffer,
// and an optional crash path for core-dump testing.
#include <cstdint>
#include <cstdio>
#include <cstring>

struct Header {
    uint32_t length;
    uint32_t magic;
};

int g_connection_count = 17;          // 0x11 once output-radix is 16
char g_hostname[32] = "prod-db-01";
static int g_retries = 3;
uint64_t g_table[4] = {0x1111111111111111ULL, 0x2222222222222222ULL,
                       0x3333333333333333ULL, 0x4444444444444444ULL};

int parse_header(const char *raw, int limit, Header *out) {
    out->length = (uint32_t)strlen(raw);
    out->magic = 0xDEADBEEF;
    int scratch[6];
    int total = 0;
    for (int i = 0; i < 6; i++) {
        scratch[i] = i * i;
        total += scratch[i];
    }
    if ((int)out->length > limit) {
        return -1;
    }
    return total;
}

int handle_packet(const char *payload, int limit) {
    Header h;
    printf("handling %zu bytes\n", strlen(payload));
    fflush(stdout);
    return parse_header(payload, limit, &h);
}

int main(int argc, char **argv) {
    printf("target starting\n");
    fflush(stdout);
    int total = handle_packet("hello world", 4096);
    printf("ok total=%d globals=%d %s retries=%d\n", total, g_connection_count,
           g_hostname, g_retries);
    fflush(stdout);

    if (argc > 1 && strcmp(argv[1], "crash") == 0) {
        Header *bad = nullptr;
        bad->length = 1;  // deliberate segfault
    }
    return 0;
}
